common/argparse.cpp

```diff
@@ -0,0 +1,1362 @@
+  auto ParseStringArgValue(const Arg& arg, llvm::StringRef value) -> bool;
+  auto ParseOneOfArgValue(const Arg& arg, llvm::StringRef value) -> bool;
+  auto ParseArg(const Arg& arg, std::optional<llvm::StringRef> value,
+                bool negated_name = false) -> bool;
+
+  auto SplitValue(llvm::StringRef& unparsed_arg)
+      -> std::optional<llvm::StringRef>;
+  auto ParseLongOption(llvm::StringRef unparsed_arg) -> bool;
+  auto ParseShortOptionSeq(llvm::StringRef unparsed_arg) -> bool;
+  auto FinalizeParsedOptions() -> bool;
+
+  auto ParsePositionalArg(llvm::StringRef unparsed_arg) -> bool;
+  auto ParseSubcommand(llvm::StringRef unparsed_arg) -> bool;
+
+  auto FinalizeParse() -> ParseResult;
+
+  llvm::raw_ostream& errors_;
+  const MetaPrinter error_meta_printer_;
+
+  const Command* command_;
+
+  OptionMapT option_map_;
+  ShortOptionTableT short_option_table_;
+  SubcommandMapT subcommand_map_;
+
+  ssize_t positional_arg_index = -1;
+  bool appending_to_positional_arg = false;
+
+  ActionT arg_meta_action_;
+};
+
+void Args::Parser::PopulateMaps(const Command& command) {
+  option_map_.clear();
+  for (const auto& option : command.options) {
+    option_map_.insert({option->info.name, {option.get(), false}});
+  }
+  short_option_table_.fill(nullptr);
+  for (auto& map_entry : option_map_) {
+    const Arg* option = map_entry.second.getPointer();
+    if (option->info.short_name.empty()) {
+      continue;
+    }
+    CARBON_CHECK(option->info.short_name.size() == 1)
+        << "Short option names must have exactly one character.";
+    unsigned char short_char = option->info.short_name[0];
+    CARBON_CHECK(short_char < short_option_table_.size())
+        << "Short option name outside of the expected range.";
+    short_option_table_[short_char] = &map_entry.second;
+  }
+  subcommand_map_.clear();
+  for (const auto& subcommand : command.subcommands) {
+    subcommand_map_.insert({subcommand->info.name, subcommand.get()});
+  }
+}
+
+void Args::Parser::SetOptionDefault(const Arg& option) {
+  CARBON_CHECK(option.has_default) << "No default value available!";
+  switch (option.kind) {
+    case Arg::Kind::Flag:
+      *option.flag_storage = option.default_flag;
+      break;
+    case Arg::Kind::Integer:
+      *option.integer_storage = option.default_integer;
+      break;
+    case Arg::Kind::String:
+      *option.string_storage = option.default_string;
+      break;
+    case Arg::Kind::OneOf:
+      option.default_action(option);
+      break;
+    case Arg::Kind::MetaActionOnly:
+      CARBON_FATAL() << "Can't set a default value for a meta action!";
+    case Arg::Kind::Invalid:
+      CARBON_FATAL() << "Option configured without any action or kind!";
+  }
+}
+
+auto Args::Parser::ParseNegatedFlag(const Arg& flag,
+                                    std::optional<llvm::StringRef> value)
+    -> bool {
+  if (flag.kind != Arg::Kind::Flag) {
+    errors_ << "ERROR: Cannot use a negated flag name by prefixing it with "
```

Comment:
It looks like a fair amount of this code bails on error. Would it be fair to consider changing to `ErrorOr<Success>` instead of using errors_ directly? The Success versus MetaSuccess result could even be `ErrorOr<bool>` where the bool indicates whether the error has already been handled, potentially removing this class from streaming errors.